_posts/2020-11-23-how-fast-are-ractors.md

```diff
@@ -0,0 +1,133 @@
+
+I'm still checking this on an itty bitty instance, a Linode 4GB. This is a very prerelease feature. And based on the results I'm seeing, it's not ready for larger-scale benchmarking yet. That's okay, let's have a look anyway.
+
+There's some setup time for these benchmarks. If I run just 5 workers with 10 messages/worker plus startup and shutdown, here are the median times in seconds it takes for each benchmark: 
+
+### Initial Setup Overhead with Only 50 Total Reqs/Process
+
+<table>
+    <thead>
+        <tr>
+            <th></th><th>Messages-Only</th><th>Whole-Process</th>
+        </tr>
+    </thead>
+    <tbody>
+        <tr> <th>fork_test</th><td>0.012</td><td>0.123</td> </tr>
+        <tr> <th>thread_test</th><td>0.051</td><td>0.159</td> </tr>
+        <tr> <th>ractor_test</th><td>0.009</td><td>0.114</td> </tr>
+        <tr> <th>pipeless_ractor_test</th><td>0.009</td><td>0.109</td> </tr>
+    </tbody>
+</table>
+
+This is just the basic setup time. But if you wonder, "isn't there some fixed overhead before you're really timing anything?" &mdash; yes, yes there is. This table gives you an idea how much overhead.
+
+What's "pipeless?" In these tests, threads are coordinating directly with a queue. But there's an added bounce through the "pipe" for Ractors to allow them to finish at different speeds &mdash; that's how [Koichi's example code](https://github.com/ruby/ruby/blob/master/doc/ractor.md#worker-pool) did it. But I thought, maybe it's faster without it?
+
+I removed the pipe and it ***was*** faster. So: ["ractor_test" is similar to Koichi's design](https://github.com/noahgibbs/ractor_basic_benchmarks/blob/main/benchmarks/ractor_test.rb), while ["pipeless" doesn't use an extra coordinator Ractor](https://github.com/noahgibbs/ractor_basic_benchmarks/blob/main/benchmarks/pipeless_ractor_test.rb).
+
+## Hypotheses and Results
+
+We would expect Ractors, at best, to be faster than existing Ruby threads because they can use all the cores available. And they would be (at best) about the same speed as multiple processes, up until all memory on the machine is used. Multiprocess concurrency in Ruby does very well, as you've [probably seen on previous benchmarks](https://appfolio-engineering.squarespace.com/appfolio-engineering/2019/9/4/benchmark-results-threads-processes-and-fibers).
+
+I originally benchmarked this on a Mac and everything else beat Ractors by a mile, plus they crashed a lot. Then I switched over to Linux and got much saner results, much more like I expected. I got literally ***one*** crash in all the Linux testing for this article rather than a literally 20% crash rate. Don't use Mac yet.
```

Comment:
Just a suggestion here: 

```suggestion
I originally benchmarked this on a Mac and everything else beat Ractors by a mile, plus they crashed a lot. Then I switched over to Linux and got much saner results, much more like I expected. I got literally ***one*** crash in all the Linux testing for this article rather than a literally 20% crash rate. Quick reminder: Don't use Mac yet.
```